---

GenomicRanges_Rle_Note
---

```{r dependencies, warning=FALSE, message=FALSE}
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

## Overview

The `Biocpkg("GenomicFeatures")` package contains functionality for transcript database or *TxDb* objects.  These objects contains a coherent interface to transcripts.

## Examples

We examine a database of human transcripts.

```{r txdb}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
txdb
```

A `TxDb` object is an interface to a `SQLite` database.


**Extract basic quantities**

- `genes()`
- `transcripts()`
- `cds()`
- `exons()`
- `microRNAs()`
- `tRNAs()`
- `promoters()`

**Extract quantities and group**

- `transcriptsBy(by = c("gene", "exon", "cds"))`
- `cdsBy(by = c("tx", "gene"))`
- `exonsBy(by = c("tx", "gene"))`
- `intronsByTranscript()`
- `fiveUTRsByTranscript()`
- `threeUTRsByTranscript()`

(Note: there are grouping functions without the non-grouping function and vice versa; there is for example no `introns()` function.

**Other functions**

- `transcriptLengths()` (optionally include CDS length etc).
- `XXByOverlaps()` (select features based on overlaps with `XX` being `transcript`, `cds` or `exon`).

**Mapping between genome and transcript coordinates**

- `extractTranscriptSeqs()` (getting RNA sequencing of the transcripts).

## Gene, exons and transcripts

Let us examine genes, exons and transcripts. We focus on a single gene on chr1: DDX11L1.
```{r gr}
gr <- GRanges(seqnames = "chr1", strand = "+", ranges = IRanges(start = 11874, end = 14409))
subsetByOverlaps(genes(txdb), gr)
subsetByOverlaps(genes(txdb), gr, ignore.strand = TRUE)
```
The `genes()` output contains a single gene with these coordinates, overlapping another gene on the opposite strand.  

Note that the gene is represented as a single range; so this output tells us nothing about exons and splicing. 

There is a single identifier called `gene_id`.  The output of `txdb` contains an "Entrex Gene ID".

```{r transcripts}
subsetByOverlaps(transcripts(txdb), gr)
```
The gene has 3 transcripts; again we only have coordinates of the pre-mRNA here.  There are 3 different transcript names (`tx_name`) which are identifiers from UCSC and then we have a `TxDb` specific transcript id (`tx_id`) which is an integer.  Let's look at exons:
```{r exons}
subsetByOverlaps(exons(txdb), gr)
```
Here we get 6 exons, but no indication of which exons makes up which transcripts.  To get this, we can do
```{r exonsBy}
subsetByOverlaps(exonsBy(txdb, by = "tx"), gr)
```
Here we see the structure of the three transcripts in the form of a `GRangesList`.

Let us include the coding sequence (CDS).  
```{r cds}
subsetByOverlaps(cds(txdb), gr)
subsetByOverlaps(cdsBy(txdb, by = "tx"), gr)
```

The output of `cds()` is not very useful by itself, since each range is part of a CDS, not the entire cds.  We need to know how these ranges together form a CDS, and for that we need `cdsBy(by = "tx")`.
```{r transcriptLengths}
subset(transcriptLengths(txdb, with.cds_len = TRUE), gene_id == "100287102")
```
(here we subset a `data.frame`).

Note: as an example of terminology mixup, consider that the output of `transcripts()` are coordinates for the unspliced transcript, whereas `extractTranscriptSeqs()` is the RNA sequence of the spliced transcripts.
