---
Di Zhen

GenomicRanges_GRanges_Usage_Note
---

```{r}
library(GenomicRanges)
```

## DataFrame

The `DataFrame` class allows columns of any class, provided a number of required methods are supported. For example, `DataFrame` can have `IRanges` as columns: 

```{r DataFrame}
ir <- IRanges(start = 1:2, width = 3)
ir
df1 <- DataFrame(iranges = ir)
df1
df1$iranges
df2 <- data.frame(iranges = ir)
df2
```

In the `data.frame` case, the `IRanges` gives rise to 4 columns, whereas it is a single column when a `DataFrame` is used.

## GRanges, metadata

The formal way to access and set `GRanges` (unlike `IRanges`) associated metadata is through `values` or `elementMetadata` or `mcols`, like

```{r GRanges}
gr <- GRanges(seqnames = "chr1", strand = c("+", "-", "+"),
              ranges = IRanges(start = c(1,3,5), width = 3))
gr
values(gr) <- DataFrame(score = c(0.1, 0.5, 0.3))
gr
``` 

Can also set and access metadata is through the `$` operator

```{r grdollar}
gr$score
gr$score2 = gr$score * 0.2
gr
```

## findOverlaps

`findOverlaps` works exactly as for `IRanges`.  But the `strand` information can be confusing.  Let us make an example

```{r findOverlaps_setup}
gr2 <- GRanges(seqnames = c("chr1", "chr2", "chr1"), strand = "*",
               ranges = IRanges(start = c(1, 3, 5), width = 3))
gr2
gr
```

The `ranges` in the two `GRanges` object are the same coordinates with different `seqnames` and `strand`.

```{r findOverlaps}
findOverlaps(gr, gr2)
```

The `*` strand overlaps both `+` and `-`.  There is an argument `ignore.strand` to `findOverlaps` which will ignore the strand information (so `+` overlaps `-`).

## subsetByOverlaps

`subsetByOverlaps`: To select only certain ranges from a `GRanges` which overlap something else.

```{r subsetByOverlaps}
subsetByOverlaps(gr, gr2)
```

## makeGRangesFromDataFrame

The `makeGRangesFromDataFrame` converts `data.frame` into a `GRanges`. `keep.extra.columns`: whether you want to keep any additional columns.

```{r makeGRangesFromDataFrame}
df <- data.frame(chr = "chr1", start = 1:3, end = 4:6, score = 7:9)
df
makeGRangesFromDataFrame(df)
makeGRangesFromDataFrame(df, keep.extra.columns = TRUE)
```

## Biology usecases


### Biology usecase I

Suppose we want to identify transcription factor (TF) binding sites that overlaps known SNPs.

Input objects are  
**snps**: a `GRanges` (of width 1)  
**TF**: a `GRanges`

pseudocode:
```{r usecaseI, eval=FALSE}
findOverlaps(snps, TF)
``` 

## Biology usecase II

Suppose we have a set of differentially methylation regions (DMRs) (think genomic regions) and a set of CpG Islands and we want to find all DMRs within 10kb of a CpG Island.

Input objects are  
**dmrs**: a `GRanges`  
**islands**: a `GRanges`

pseudocode:
```{r usecaseII, eval=FALSE}
big_islands <- resize(islands, width = 20000 + width(islands), fix = "center")
findOverlaps(dmrs, big_islands)
``` 

