---
title: "Project 5. Genomic Data Science for Epigenetics"
output:
  html_document:
    df_print: paged
    toc: yes
  word_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE,
                      fig.height = 3.2,
                      fig.width = 6.5)
```

## The Outline for Project 5.

Project 5 aims to discover a set of genomic features that could explain and predict patterns of epigenetic status on the genome. Firstly, we will try to apply basic Bioconductor tools to extract the genomic land markers from the gene annotations. Then, we will analyze a published epigenetic modification data using the extracted annotations. Explicitly, we will investigate its consensus motifs, exon lengths, as well as its meta distributions on genes. In the end, we will try to use machine learning to develop a prediction tool for the epigenetic modification, and see if we can integrate the genome topologies discovered from the EDA to enhance its prediction performance.

This project tries to reproduce the key findings in the following papers, please check them if you need more background information:

*1.Dominissini D, Moshitch-Moshkovitz S, Schwartz S, Salmon-Divon M, Ungar L, Osenberg S, Cesarkas K, Jacob-Hirsch J, Amariglio N, Kupiec M: Topology of the human and mouse m6A RNA methylomes revealed by m6A-seq. Nature 2012, 485(7397):201-206.*

*2.Meyer KD, Yogesh S, Paul Z, Olivier E, Mason CE, Jaffrey SR: Comprehensive analysis of mRNA methylation reveals enrichment in 3' UTRs and near stop codons. Cell 2012, 149(7):1635–1646.*

*3.Chen K, Wei Z, Zhang Q, Wu X, Rong R, Lu Z, Su J, de Magalhães JP, Rigden DJ, Meng J: WHISTLE: a high-accuracy map of the human N6-methyladenosine (m6A) epitranscriptome predicted using a machine learning approach. Nucleic Acids Research 2019.*

## 1. Extract Transcriptomic Landmarkers

First, our goal is to extract some important genomic features for the later analysis. Retrieve the 4 transcriptomic landmarks shown below from the `TxDb` package of hg19. 

- Transcription Start Sites (TSS).
- Transcription Ending Sites (TES).
- Stop Codon.
- Start Codon.

To achieve this, we need to rely on the combination of a set of "intra-range methods". The extracted landmarks should be `GRanges` object with width = 1. Please store the results into the variables of `TSS`, `TES`, `Start_codon`, and `Stop_codon`, respectively. 

Hint: `transcripts()` can extract the `GRanges` of full transcript (with introns) from `TxDb` object, and `cdsBy()` can retrieve a `GRangesList` that contains exons of CDS for each transcript / genes.  Please try to understand these data structures, and be careful that the `cdsBy()` returns multiple exons of a CDS, not the full CDS with introns.

```{r}
## === Hint code, fill the "___" ============= ##
# library(TxDb.___.UCSC.___.knownGene)
# txdb_hg19 <- TxDb.___.UCSC.___.knownGene
# TSS <- ___(transcripts(txdb_hg19), width = ___, fix = ___)
# TES <- ___(transcripts(txdb_hg19), width = ___, fix = ___)
# Start_codon <- ___(unlist(___(cdsBy(txdb_hg19, by = ___))), width = ___, fix = ___)
# Stop_codon <- ___(unlist(___(cdsBy(txdb_hg19, by = ___))),width = ___, fix = ___)
# #Some extra hints, the missing functions above are all intra-range methods
## ===== Enter your code below  =============== ##
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb_hg19 <- TxDb.Hsapiens.UCSC.hg19.knownGene
TSS <- resize(transcripts(txdb_hg19), width = 1, fix = "start")
TES <- resize(transcripts(txdb_hg19), width = 1, fix = "end")
Start_codon <- resize(unlist(range(cdsBy(txdb_hg19, by = "tx"))), width = 3, fix = "start")
Stop_codon <- resize(unlist(range(cdsBy(txdb_hg19, by = "tx"))), width = 3, fix = "end")
```

- SAQ1: How many unique coordinates for TSS, TES, Start codons, and Stop codons exist on the genome assembly hg19? Is there more alternative usage for the start codons or stop codons? How about the alternative usage for the transcription starts and ends?

- Answer: 82960 unique coordinates for TSS and TES. 63691 unique coordinates for start codons nad stop codons. Alternate start codons (non-AUG) are very rare in eukaryotic genomes. Naturally occurring non-AUG start codons have been reported for some cellular mRNAs. Seven out of the nine possible single-nucleotide substitutions at the AUG start codon of dihydrofolate reductase were functional as translation start sites in mammalian cells. Alternative stop codon includes AGA and AGG. Alternative transcription start and termination sites are the principal drivers of transcript isoform diversity across tissues, and may underlie the majority of cell type specific proteomes and functions.

### Examine the Sequence Around the Start and Stop Codons

In this step, we will extract the sequences of the 3 bases windows centered by the Start and Stop condons, then, we will plot the sequence logo around the stop and start codons.

A sequence logo is a graphical representation of the conservation on amino acid or nucleic acid fragments. In R, the sequence logo can be plotted using function `ggseqlogo()` in the package `ggseqlogo`. The input of `ggseqlogo()` can be a string vector of nucleotide sequences that have the same string lengths. Please set the method = "prob", so that it will display the proportion of bases at each nucleotide position.

```{r}
## === Hint code, fill the "___" ============= ##
#library(ggseqlogo)
#library(BSgenome.___.UCSC.___)
#ggseqlogo(as.vector( ___( ___(Hsapiens,Start_codon+___) ) ), method = "prob")
#ggseqlogo(as.vector( ___( ___(Hsapiens,Stop_codon+___) ) ), method = "prob")
## ===== Enter your code below  =============== ##
library(ggseqlogo)
library(BSgenome.Hsapiens.UCSC.hg19)
ggseqlogo(as.vector( as.character( getSeq(Hsapiens,Start_codon) ) ), method = "prob")
ggseqlogo(as.vector( as.character( getSeq(Hsapiens,Stop_codon) ) ), method = "prob")
```

Afterward, check if the generated sequence logos being mostly consistent with the codon sequences in the common knowledge.

## 2. Identify Sequence Motif of an Epigenetic Modification

Subsequently, we will begin to analyze an epigenetic dataset downloaded from GEO. There is a BED file named `GSE63753.bed` under the project directory. The BED file was downloaded from the supplementary data of the GEO dataset [GSE63753](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63753). 

The GEO data is published by a study that detect the m6A modifications using miCLIP. miCLIP is a sequencing technology that can detect nucleotide modifications on the transcriptome at a single-base resolution. The file contains one of the processed miCLIP sample in the study, which reveals a single-based methylome of m6A in the HEK293 cell line.

Use the function `rtraklayer::import()` to load the BED file `GSE63753.bed` into R, and save the GRanges in a variable named `m6A`. Subsequently, extract the sequences under the 5bp window centered by the m6A modification. Using the same method of the previous step, plot the frequency of the nucleotides for the pentamer sequences.

```{r}
## === Hint code, fill the "___" ============= ##
# m6A <- rtracklayer::import(___)
# sequence <- ___
# ggseqlogo( sequence, ___ )
# ===== Enter your code below  =============== ##
m6A <- rtracklayer::import("GSE63753.bed")
sequence <- as.vector(as.character(getSeq(Hsapiens,m6A+2)))
ggseqlogo( sequence, method = "prob" )
```

- SAQ2: Check the IUPAC nucleotide ambiguity codes defined in the variable `IUPAC_CODE_MAP`. Then, describe a consensus motif using 5 letters, such that the motif you named can capture >70% of the sequences in the sequence logo plot. The answer cannot be too vague (such as NNNNN), the scope of your motif should be as narrow as possible in terms of the percentage among all possible base combinations of a pentamer. Please briefly explain your reasoning.

- Answer: DDACH. "D" meams "AGT", "H" means "ACT". As shown in the logo above, each base in "DDACH" captures >70% of the base in 5bp-window m6A. The result of evaluation below shows that "DDACH" matches 84.3% of the m6A, and it is also not too vague.

You can evaluate the proportion of the m6A matching with:
```{r}
IUPAC_CODE_MAP
mean(vcountPattern("DDACH", DNAStringSet( Views( Hsapiens, m6A + 2) ), fixed = F)) #Change some bases in GGACA into ambiguity code so that it can include more motifs
```

## 3. Plot the Length of Exons

Next, we will explore the correlation between exon lengths and m6A modification. One of the EDA approaches is to plot the exon length distribution for exons containing or not containing the m6A sites.

Fill in the code below so that we will create 2 vectors of the same length. The first vector named `length_exon` contains length of each exon on hg19, while the second vector named `overlap_m6A` is a dummy variable (logic), and it will be true if the exon contains m6A. Then, run the code afterwards to plot the densities of exon lengths stratified by the overlapping status with the modification.

```{r}
## === Hint code, fill the "___" ============= ##
#ex_hg19 <- ___(___) #extract exons of hg19 using exons()
#length_exon <- ___(___) #extract the lengths for each of the exons
#overlap_m6A <- ___%___%___ #retrieve a dummy variable, true if exon overlapps with m6A
# ===== Enter your code below  =============== ##
ex_hg19 <- exons(txdb_hg19) #extract exons of hg19 using exons()
length_exon <- width(ex_hg19) #extract the lengths for each of the exons
overlap_m6A <- ex_hg19%over%m6A #retrieve a dummy variable, true if exon overlapps with m6A
# ===== Your code is finished ================ ##
library(ggplot2)
library(ggsci)
ggplot(data.frame(length_exon,overlap_m6A)) +
                   geom_density(aes(x = log(length_exon),
                                    colour = overlap_m6A,
                                    fill = overlap_m6A), alpha = 0.5) + 
                   theme_classic() + scale_color_npg()
```

- SAQ3: Please interpret the density plot. What is the difference between the exons containing the modification and the exons without the modification? Are we more likely to observe m6A on long exons? If there is a boundary of exon length to classify the exons containing or not containing the m6A, what boundary would you choose?

- Answer: This plot shows the density against the log of exon length of exons with and without modification. The mean of exon length of exons with modification is larger than it of exons without modification. The standard deviation of exon length of exons with modification is larger than it of exons without modification. Yes, we are more likely to observe m6A on long exons. I would choose the log of exon length to be 5.5 as the boundary to classify the exons. 

## 4. Plot the Topology (Distribution) of Markers on Genes

Next, we will draw a meta-gene plot. Specifically, the figure summarizes the spatial distribution of genomic features on transcript coordinates relative to start and the stop codons. You could refer to Figure 5D in Meyer KD.et.al Nature (2012) as an example.

The information we need to draw this graph is not complex: the only required values are the relative positions of each site on the 5'UTR, CDS, and 3'UTR. After calculating the relative positions, all we have to do is to draw 3 histograms of the relative position of each region, and the final distribution is just a combination of these 3 histograms.

In the following code chunk, you need to create a function called `relative_pos_on_region()`. The function will return the relative position of its input GRanges (parameter `x`) on the GRangesList (parameter `region`). The relative position on a region is defined as the distance toward the 5' Start of the region divided by the full length of the region. 

For example, if a site is located on the 200bp downstream of the start of a 5'UTR, and the length of the 5'UTR is 1000, then its relative position on 5'UTR is 200/1000 = 0.2. If some entries of x are not in the region, the function should omit those entries and only return the relative positions for entries mapped to the region. 

```{r}
## === Hint code, fill the "___" ============= ##
#relative_pos_on_region <- function(x, region){
#region_map <- map___Transcripts(___, ___)
#region_width <- sum(___(___))[region_map$___]
#start_on_region <- ___(region_map)
#return(___/___)
#}
# ===== Enter your code below  =============== ##
relative_pos_on_region <- function(x, region){
  region_map <- mapToTranscripts(x, region)
  region_width <- sum(width(region))[region_map$transcriptsHits] 
  start_on_region <- start(region_map)
  return(start_on_region/region_width)
}
# ===== Your code is finished ================ ##
plot_tx <- function(marker,utr5_gr,cds_gr,utr3_gr,marker_name){
  utr5_pos <- relative_pos_on_region(marker,utr5_gr)
  cds_pos <- relative_pos_on_region(marker,cds_gr)
  utr3_pos <- relative_pos_on_region(marker,utr3_gr)

  pldf <- data.frame(relative_pos = c(utr5_pos, cds_pos, utr3_pos),
                     tx_region = rep(c("5'UTR","CDS","3'UTR"),
                                     c(length(utr5_pos),length(cds_pos),length(utr3_pos)))
             )

  pldf$tx_region = factor(pldf$tx_region,levels = c("5'UTR","CDS","3'UTR"))

  ggplot(pldf) +
    geom_histogram(aes(x=relative_pos),bins = 50,
                   colour = "black", fill = "black") +
    facet_wrap(~tx_region) +
    theme_classic() +
    labs(title = marker_name, x = "Relative Position")
} #What this function do is just organize the data and plot the histogram.

#Extract the regions for 5'UTR, CDS, and 3'UTR
UTR5 <- fiveUTRsByTranscript(txdb_hg19) 
CDS <- cdsBy(txdb_hg19, by = "tx")
UTR3 <- threeUTRsByTranscript(txdb_hg19)

#Then we will generate a serious of plot for the topology of previously extracted tx landmarkers
plot_tx(TSS,UTR5,CDS,UTR3,"TSS") # 
plot_tx(TES,UTR5,CDS,UTR3,"TES")
plot_tx(Stop_codon,UTR5,CDS,UTR3,"Stop Codon")
plot_tx(Start_codon,UTR5,CDS,UTR3,"Start Codon") #
plot_tx(m6A,UTR5,CDS,UTR3,"m6A") #

```

- SAQ4: Please interpret the computed topology distributions for m6A. Which transcript regions are more likely to contain the modification sites? Are the relative positions on the regions important for m6A? Around which transcript landmarks (e.x. TSS, start codon) is the modification most enriched? Please explain your reasoning.

- Answer: There is few m6A in 5'UTR and more m6A in CDS and 3'UTR. m6A is mainly concentrated at the end of CDS and at the begin of 3'UTR, which is the position of stop codon. The relative positions on the regions are important for m6A. Around stop codon is the modification most enriched. From this enriched position, we can deduce the function of m6A modification. m6A affects RNA transcription, processing, translation and metabolism.

## 5. Engineering Genomic Features for Site Prediction of Epigentic Markers

In project 3, we introduced the unsupervised learning methods for the prediction of tissue labels from gene expression data. Now, we want to use the machine learning model again to predict the modification status on consensus motifs. In this case, we want to use a supervised learning methods, which means that the target of the prediction is provided to the model during the model training process. 

Our training data is stored in the `mcols()` of `m6A_ml`, which contains 1000 training instances for both the positive and negative data. The metadata is a `DataFrame` that has columns for target and sequence-based features. The sequence-based features are 30 variables extracted only from the nucleotide sequences surrounding the modification sites, and their calculation is NOT dependent on the gene annotations (from only BSgenome but not from TxDb). What you need to do is to create the following annotation based genomic features as the additional metadata columns in `m6A_ml`: 

- `UTR5`: a logical variable indicating the site overlapping with 5'UTR.
- `CDS`: a logical variable indicating the site overlapping with CDS.
- `UTR3`: a logical variable indicating the site overlapping with 3'UTR.
- `long_exon`: a logical variable for sites overlapping with long exons (exon length > XXX, XXX is a boundary number choosen by you).
- `Stop_codon`: a logical variable for sites overlapping with the XXX bp centered by a stop codon. (XXX is also choosen by you)
- `UTR3_pos`: a real number value for the relative position of the site on 3'UTR (0 if not on 3'UTR).
- `CDS_pos`: a real number value for the relative position of the site on CDS (0 if not on CDS).

Please fill the following code chunk to add the above-mentioned genomic features into the prediction models.

After adding the features, we will run a high-level package to automatically build 4 prediction models and report their performance using 5 folds cross-validation.  The 4 models are a combination of 2 ML algorithms (SVM and RandomForest) and 2 feature types (sequence feature along and sequence + genomic features).

You will get full score for this part if the genomic features you engineered can lead to an AUROC of more than 0.70 in any one of the ML algorithms.

```{r}
m6A_ml <- readRDS("m6A_ml.rds") #m6A_ml is a GRanges, its metadata columns are features used in the prediction model

## === Hint code, fill the "___" ============= ##
#m6A_ml$UTR5 <- m6A_ml %___% ___ #add a dummy feature for 5'UTR
#m6A_ml$CDS <- m6A_ml ___ ___ #add a dummy feature for 5'CDS
#m6A_ml$UTR3 <- m6A_ml ___ ___ #add a dummy feature for 3'UTR
#m6A_ml$long_exon <- m6A_ml ___ ___ #add a dummy feature for long exon (> a boundary number)
#m6A_ml$Stop_codon <- m6A_ml ___ ___ #add a dummy feature for stop codon flanked by a choosen number
#m6A_ml$UTR3_pos <- WhistleR:::relative_pos_map(m6A_ml, UTR3, 0, F) #a feature for relative position on UTR3
#m6A_ml$CDS_pos <- WhistleR:::relative_pos_map(m6A_ml, CDS, 0, F) #a feature for relative position on CDS
# ===== Enter your code below  =============== ##
m6A_ml$UTR5 <- m6A_ml %over% UTR5 #add a dummy feature for 5'UTR
m6A_ml$CDS <- m6A_ml %over% CDS #add a dummy feature for 5'CDS
m6A_ml$UTR3 <- m6A_ml %over% UTR3 #add a dummy feature for 3'UTR
# exstx_hg19 <- unlist(exonsBy(txdb_hg19, by = "tx"))
m6A_ml$long_exon <- m6A_ml %over% ex_hg19[length_exon > 400] #add a dummy feature for long exon (> a boundary number)
m6A_ml$Stop_codon <- m6A_ml %over% (Stop_codon + 100) #add a dummy feature for stop codon flanked by a choosen number
m6A_ml$UTR3_pos <- WhistleR:::relative_pos_map(m6A_ml, UTR3, 0, F) #a feature for relative position on UTR3
m6A_ml$CDS_pos <- WhistleR:::relative_pos_map(m6A_ml, CDS, 0, F) #a feature for relative position on CDS

# ===== Your code is finished ================ ##
library(perflite)
library(knitr)
set.seed(102)

perf_results <- performance_class(
  y = list(
    target_1 = as.factor(m6A_ml$m6A),
    target_2 = as.factor(m6A_ml$m6A)
  ), #list of response vectors
  X = list(
    sequence_feature = data.frame( mcols(m6A_ml)[,2:31] ),
    add_genomic_feature =  data.frame( mcols(m6A_ml)[,-1] ) 
  ), #list of feature matrixes
  k = 5, #number of folds in cross validation
  p = 1, #number of parallel computation
  cv_f = c(svm = svm_class,
           randomForest = randomForest_class)  #list of classifier functions.
)

kable(perf_results[[1]], 
       caption = names(perf_results)[1]) 

kable(perf_results[[2]],
       caption = names(perf_results)[2]) 
```

- SAQ5: For the prediction made by SVM, after adding genomic features (annotation-based), by how much is the AUROC improved compared with only using the sequence features? Can you explain such an improvement? If an annotation-based feature is important at improving the prediction performance, can we infer that the annotation is biologically/scientifically significant for the predicted epigenetic marker? Please explain your reasoning.

- Answer: Using SVM, the AUROC of sequence features is 0.6053, and the AUROC after adding genomic feature is 0.7068. Using RandomForest, the AUROC of sequence features is 0.6077, and the AUROC after adding genomic features is 0.7028. The AUROC has improved about 1/6 from roughly 60% to 70%. This is because genomic features are important and can contribute to the performance of prediction model. After combining genomic features and sequence features, the features in total improve the performance of the model a lot. Yes, we can infer that the annotation is biologically significant for the predicted epigenetic marker. Because those genetic features can be used by the trained model as important measures to split the data into two groups (with and without m6A). This means the genomic features are more or less significantly different between the positive group and the negative group. In biology, if the genomic features are different between the groups with and without m6A, those genomic features are more likely associated with the existance of m6A. Thus, these features are helpful in predicting m6A. So we can infer that the annotation is biologically/scientifically significant for the predicted epigenetic marker.

## Session Info

```{r}
sessionInfo()
```


